src/output.c

```diff
@@ -768,3 +778,55 @@ handle_output_power_manager_set_mode(struct wl_listener *listener, void *data)
 		break;
 	}
 }
+
+void
+virtual_output_add(struct server *server, const char *output_name)
+{
+	if (output_name) {
+		/*
+		 * Prevent creating outputs with the same name
+		 */
+		struct output *output;
+		wl_list_for_each(output, &server->outputs, link) {
+			if (wlr_output_is_headless(output->wlr_output)) {
+				if (!strcmp(output->wlr_output->name, output_name)) {
+					wlr_log(WLR_DEBUG,
+					"refusing to create virtual output with duplicate name");
+					return;
+				}
+			}
+		}
+		strncpy(server->headless.pending_output_name, output_name,
+				sizeof(server->headless.pending_output_name));
+	} else {
+		server->headless.pending_output_name[0] = '\0';
+	}
+	wlr_headless_add_output(server->headless.backend, 1920, 1080);
+}
+
+void
+virtual_output_remove(struct server *server, const char *output_name)
+{
+	struct output *output;
+	wl_list_for_each(output, &server->outputs, link) {
+		if (wlr_output_is_headless(output->wlr_output)) {
+			if (output_name) {
+				/*
+				 * Given virtual output name, find and destroy virtual output by
+				 * that name.
+				 */
+				if (!strcmp(output->wlr_output->name, output_name)) {
+					wlr_output_destroy(output->wlr_output);
+					return;
+				}
+			} else {
+				/*
+				 * When virtual output name was no supplied by user, simply
+				 * destroy the first virtual output found.
+				 */
+				wlr_output_destroy(output->wlr_output);
```

Comment:
Outputs are added to list with `wl_list_insert` (rather than `wl_list_append`) so you'd remove the _last_ virtual output.

Reply:
Indeed. But note that it says "the first virtual output **found**"

Reply:
I'm easy - not really a merge-stopper.

I think that changing the man-page to say "the last virtual output added" is a very good idea though. Explicit is better than implicit and all that :smile:

Reply:
Good point!